File: October/1/subsets2.py
# LeetCode 90: Subsets II
# Explanation:
# 1. We need all possible subsets of nums (the power set).
# 2. Duplicates must not appear in the result.
# 3. Sort nums first to group duplicates together.
# 4. Use backtracking:
#    - Add current subset to results.
#    - At each step, skip duplicate elements by checking nums[i] == nums[i-1].
# Time Complexity: O(2^n)
# Space Complexity: O(2^n) (for storing results)

class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        
        def backtrack(start, path):
            res.append(path[:])
            for i in range(start, len(nums)):
                if i > start and nums[i] == nums[i - 1]:
                    continue
                path.append(nums[i])
                backtrack(i + 1, path)
                path.pop()
        
        backtrack(0, [])
        return res

File: October/1/subsets2.cpp
// LeetCode 90: Subsets II
// Explanation:
// 1. Sort nums to handle duplicates.
// 2. Use backtracking, skipping duplicates (nums[i] == nums[i-1]).
// 3. Generate all unique subsets.
// Time Complexity: O(2^n)
// Space Complexity: O(2^n)

#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> path;
        sort(nums.begin(), nums.end());
        backtrack(nums, 0, path, res);
        return res;
    }
    
    void backtrack(vector<int>& nums, int start, vector<int>& path, vector<vector<int>>& res) {
        res.push_back(path);
        for (int i = start; i < nums.size(); i++) {
            if (i > start && nums[i] == nums[i-1]) continue;
            path.push_back(nums[i]);
            backtrack(nums, i+1, path, res);
            path.pop_back();
        }
    }
};

